### PR TITLE
refactor(Core/SendListInventory): Add multivendor function

### PR DESCRIPTION
**Pull request by illunix**
## HOW IT WORKS?
Creature send vendor window to player of selected npc_vendor entry

## HOW TO TEST THE CHANGES:

1. Recompile core with this script 
```cpp
class npc_test_vendor : public CreatureScript
{
public:
    npc_test_vendor() : CreatureScript("npc_test_vendor") { }

    bool OnGossipHello(Player* player, Creature* creature) override
    {
        AddGossipItemFor(player, GOSSIP_ICON_INTERACT_1, "Armor", GOSSIP_SENDER_MAIN + 1, GOSSIP_ACTION_INFO_DEF);
        SendGossipMenuFor(player, DEFAULT_GOSSIP_MESSAGE, creature->GetGUID());
        return true;
    }

    bool OnGossipSelect(Player* player, Creature* creature, uint32 uiSender, uint32 action) override
    {
        switch (uiSender)
        {
        case GOSSIP_SENDER_MAIN + 1:
            player->GetSession()->SendListInventory(creature->GetGUID(), 8888);
            return true;
        }
    }
};

void AddSC_npc_test_vendor()
{
    new npc_test_vendor();
}
```
1. Create npc with `129` flag, `134217728` (FORCE_GOSSIP) type_flags and script name `npc_test_vendor`
2. Add items to `8888` npc
3. Clear client cache then open game
4. Spawn npc with `npc_test_vendor` script and click the first option, if you get vendor window with items you added to `8888` npc everything works

### CODE REVIEW

**Comment by Kitzunu** [comment]
1. Please fix build
2. Could you provide a screenshot?
3. This sounds more like a module to me?

**Comment by Illunix** [comment]
1. Working on windows but yeah I'll fix
2. Yeah of course in a second
3. Nah, it's just a core edit what allow to do multivendor npc
@Kitzunu

**Comment by Illunix** [comment]
UdHr0XAYyt
Of course player can add more options, it works for sql script too.

**Comment by Illunix** [comment]
https://github.com/azerothcore/azerothcore-wotlk/pull/3157/commits/3db1070cf72eb42ed09b5c0fef643544a169e1e8 Forgot to change it

**Comment by Kitzunu** [comment]
Please fix the build and the Codacy Warnings https://app.codacy.com/manual/FrancescoBorzi/azerothcore-wotlk/pullRequest?prid=5746274

**Comment by Helias** [comment]
I was planning to do this after a user asked in Discord this functionality.
But I was thinking to make this "optional" by a conf, but in this way should be **better** ;-)

**Comment by Kitzunu** [comment]
https://app.codacy.com/manual/FrancescoBorzi/azerothcore-wotlk/pullRequest?prid=5746274

**Comment by Illunix** [comment]
> https://app.codacy.com/manual/FrancescoBorzi/azerothcore-wotlk/pullRequest?prid=5746274

I didn't edit this, idk why it showing up there @Kitzunu

File: src/server/scripts/Commands/cs_npc.cpp [hunk 1]
```diff
@@ -261,51 +261,48 @@ class npc_commandscript : public CommandScript
...
-        uint32 maxcount = 0;
+        auto fmaxcount = strtok(nullptr, " ");                    //add maxcount, default: 0
+        auto maxcount = 0;
         if (fmaxcount)
             maxcount = atol(fmaxcount);
 
-        char* fincrtime = strtok(nullptr, " ");                    //add incrtime, default: 0
-        uint32 incrtime = 0;
+        auto fincrtime = strtok(nullptr, " ");                    //add incrtime, default: 0
+        auto incrtime = 0;
```
Comment (Kitzunu) [inline src/server/scripts/Commands/cs_npc.cpp:282]: Yes you did

File: src/server/scripts/Commands/cs_npc.cpp [hunk 2]
```diff
@@ -261,51 +261,48 @@ class npc_commandscript : public CommandScript
...
+        auto itemId = atol(pitem);
+        if (itemId <= 0)
             return false;
 
-        uint32 itemId = item_int;
-
-        char* fmaxcount = strtok(nullptr, " ");                    //add maxcount, default: 0
-        uint32 maxcount = 0;
+        auto fmaxcount = strtok(nullptr, " ");                    //add maxcount, default: 0
+        auto maxcount = 0;
```
Comment (Kitzunu) [inline src/server/scripts/Commands/cs_npc.cpp:277]: you did

File: src/server/scripts/Commands/cs_npc.cpp [hunk 3]
```diff
@@ -261,51 +261,48 @@ class npc_commandscript : public CommandScript
...
+        auto fincrtime = strtok(nullptr, " ");                    //add incrtime, default: 0
+        auto incrtime = 0;
         if (fincrtime)
             incrtime = atol(fincrtime);
 
-        char* fextendedcost = strtok(nullptr, " ");                //add ExtendedCost, default: 0
-        uint32 extendedcost = fextendedcost ? atol(fextendedcost) : 0;
-        Creature* vendor = handler->getSelectedCreature();
+        auto fextendedcost = strtok(nullptr, " ");                //add ExtendedCost, default: 0
+        auto extendedcost = fextendedcost ? atol(fextendedcost) : 0;
```
Comment (Kitzunu) [inline src/server/scripts/Commands/cs_npc.cpp:287]: here too

File: src/server/scripts/Commands/cs_npc.cpp [hunk 4]
```diff
@@ -261,51 +261,48 @@ class npc_commandscript : public CommandScript
...
             handler->SendSysMessage(LANG_SELECT_CREATURE);
             handler->SetSentErrorMessage(true);
             return false;
         }
 
-        uint32 vendor_entry = vendor->GetEntry();
-
-        if (!sObjectMgr->IsVendorItemValid(vendor_entry, itemId, maxcount, incrtime, extendedcost, handler->GetSession()->GetPlayer()))
+        auto addMulti = strtok(NULL, " ");
+        auto vendorEntry = addMulti ? handler->GetSession()->GetCurrentVendorEntry() : 0;
```
Comment (Kitzunu) [inline src/server/scripts/Commands/cs_npc.cpp:297]: and here

**Comment by Kitzunu** [comment]
For me to be allowed to merge a PR there are some things that must pass:

1. All Clang builds without module must pass.
2. The docker build must pass
3. Codacy must pass
4. The PR must be tested by someone other than the one who created the PR.

That is why it is important to fix the Codacy issue

**Comment by Shard-MW** [comment]
It should be a module for me
What's the difference between this PR and Rochet's multivendor scriptb?

**Comment by Helias** [comment]
I don't think there is much difference.
A module should not easy to implement for this, I was thinking about a "conf" to make it "optional" but I think that this does not break anything, just add an "optional" feature, if you want to use it you can, otherwise you don't and the core do not change the behavior of the vendors.

**Comment by Illunix** [comment]
> I don't think there is much difference.
> A module should not easy to implement for this, I was thinking about a "conf" to make it "optional" but I think that this does not break anything, just add an "optional" feature, if you want to use it you can, otherwise you don't and the core do not change the behavior of the vendors.

True, that's why I done this

**Comment by Helias** [comment]
@Illunix, remove the variables that codacy and @Kitzunu  reported to you, they are not used.

**Comment by Illunix** [comment]
All variables what codacy report is used [here](https://github.com/azerothcore/azerothcore-wotlk/blob/master/src/server/scripts/Commands/cs_npc.cpp#L306
) So idk why I should delete them? Or what I should do with them?

**Comment by Illunix** [comment]
> All of these lines that you changed or added comes up as issues in Codacy https://app.codacy.com/manual/FrancescoBorzi/azerothcore-wotlk/pullRequest?prid=5746274

I just changed variables to ``auto``

**Comment by IntelligentQuantum** [comment]
@Illunix A simple question

Isn't this better?
https://github.com/IntelligentQuantum/azerothcore-wotlk/commit/fa0eb7e46a277662425dfe3013d5eb8a1ed71e3a

**Comment by Illunix** [comment]
> @Illunix A simple question
> 
> Isn't this better?
> [IntelligentQuantum@fa0eb7e](https://github.com/IntelligentQuantum/azerothcore-wotlk/commit/fa0eb7e46a277662425dfe3013d5eb8a1ed71e3a)

Is that work?

**Comment by IntelligentQuantum** [comment]
> > @Illunix A simple question
> > Isn't this better?
> > [IntelligentQuantum@fa0eb7e](https://github.com/IntelligentQuantum/azerothcore-wotlk/commit/fa0eb7e46a277662425dfe3013d5eb8a1ed71e3a)
> 
> Is that work?

works for me.

**Comment by disclosurez** [comment]
isn't  alot more code needed? this has always been a patch
https://github.com/disclosurez/Multi-Vendor-AzerothCore-Patch/blob/49e7bf6ddb799e08caa6afc2775d9eee3537c6bc/PATCH/MultiVendorAzerothCorebyPoszer.patch

**Comment by Helias** [comment]
I think that we could use the Quantum branch for this multi-vendor feature, if he already used it for long time can be safer with that.
What do you think @illunix @IntelligentQuantum ?

**Comment by illunix** [comment]
Yeah of course

**Comment by IntelligentQuantum** [comment]
@Helias I used this for 7 months with 200 players, I had no problem

**Comment by FrancescoBorzi** [comment]
is this still WIP ?

**Comment by Kitzunu** [comment]
Well yes, seeing as the codacy isn't resolved yet :)

**Comment by Helias** [comment]
@IntelligentQuantum if you can open a PR with your commits, we can close this and redirect this thread there

**Comment by IntelligentQuantum** [comment]
https://github.com/azerothcore/azerothcore-wotlk/pull/3172

**Pull request closed without merging**